rustls/src/server/tls12.rs

```diff
@@ -661,12 +661,20 @@ impl hs::State for ExpectCertificateVerify {
                 HandshakeType::CertificateVerify,
                 HandshakePayload::CertificateVerify
             )?;
-            let handshake_msgs = self.transcript.take_handshake_buf();
-            let certs = &self.client_cert;
 
-            self.config
-                .verifier
-                .verify_tls12_signature(&handshake_msgs, &certs[0], sig)
+            match self.transcript.take_handshake_buf() {
+                Some(msgs) => {
+                    let certs = &self.client_cert;
+                    self.config
+                        .verifier
+                        .verify_tls12_signature(&msgs, &certs[0], sig)
+                }
+                None => {
+                    cx.common
+                        .send_fatal_alert(AlertDescription::AccessDenied);
+                    Err(Error::General("client authentication not set up".into()))
```

Comment:
if this is reachable, can it have a test?

Reply:
I don't think this is actually reachable:

* We can only get to the `ExpectCertificateVerify` state if `ExpectClientKx::client_cert` is `Some`
* An `ExpectClientKx` state with `client_cert` set to `Some` is only created by `ExpectCertificate`
* In `ExpectCertificate`, we might abandon client auth, but we set `ExpectClientKx::client_cert` to `None` if so
* `ExpectCertificate` is set up only if we're doing client auth, meaning `config.verifier.offer_client_auth()` is `true`
* In `ExpectClientHello::new()`, we initialized the `HandshakeHashBuffer` with client auth enabled under the same condition

I've added a comment with a condensed version of  this reasoning.